conn.go

```diff
@@ -28,29 +29,53 @@ func (n *NetAddr) String() string {
 	return fmt.Sprintf("relay[%s-%s]", n.Remote, n.Relay)
 }
 
+func (c *Conn) Close() error {
+	return c.stream.Reset()
```

Comment:
Maybe this is relevant here. This is how I ended up doing it in gostream :

https://github.com/hsanjuan/go-libp2p-gostream/blob/master/conn.go#L38

```go
// Close closes the connection.
// Any blocked Read or Write operations will be unblocked and return errors.
func (c *conn) Close() error {
	if err := c.s.Close(); err != nil {
		c.s.Reset()
		return err
	}
	go pnet.AwaitEOF(c.s)
	return nil
}
```

If I remember well, bluntly resetting libp2p streams on `Close()` caused errors on the other side on situations where the stream closing was supposed to be a clean operation.

Reply:
Yeah, but we need to be careful:

1. Concurrent reads aren't safe. We need to interrupt any concurrent reads (somehow) before we can do this. We _should_ be able to do this by setting a read deadline but... we'll have to make sure to fix any/all of our multiplexers/transports (deadlines currently don't interrupt).
2. We probably want to set some timeouts/deadlines.
3. We should probably call `Close()` in the background as well. That will mimic a normal `close(tcpFileDescriptor)` call (which will let the kernel flush in the background).

IMO, the _real_ issue here is that `TCPConn.Close` is actually a very fancy best-effort function that does a bunch of dirty work in the background. Ideally, even if we make `Close` close both directions, we wouldn't be _that_ sloppy.